### ISSUE TITLE
DCv4 Tracker

### ISSUE THREAD

**Issue by kum-deepak** [dc-js/dc.js#1586]
When I started on this path I did not assume that it can be completed without causing significant change in interface. However, based on current progress we can manage with limited interface changes. This leads me to think to make a release at this stage, i.e., skip any refactoring that will need interface breakage.

Let us use this task to track our progress till dcv4 public release.

Current TODO list:

- Review all outstanding ES6 tasks, close/defer as needed.
- Update jsdoc - since these are now ES6 classes, these can be cleaned up.
- Remove polyfills from web/js.
- Convert deprecated/annotated methods to ES6 style.
- Convert direct method assignments to ES6 style.
- Move all compat code to a file of its own.
- Convert samples to use newer syntax (Optional).
- Convert specs to use newer syntax (Optional).
- Write samples to show usage of modules (Optional).
- Add ES6 modules as alternate output in dist (Optional).
- Change folder layout for docs and web (Optional).

@gordonwoodhull, looking for your suggestions.

**Comment by gordonwoodhull** [comment]
Thanks @kum-deepak! This looks like a great plan! Sounds like we are really close.

The only thing I see missing from the original plan is compatibility with D3v6. But I agree, it makes sense to release this first, and we can do point releases that update to each ES6 D3 module as it comes out, since those will also break user code. See #1545. 

Maybe something like: 4.0.*  is compatible with D3v5, while 4.1 uses the latest modules instead of the D3 bundle. Suggestions welcome!

I agree on all the non optional stuff. I am curious about the jsdoc updates, could you explain?

Of the optional stuff, the two items regarding ES6 modules stand out to me. What does an ES6 module distribution look like - is it just a copy of `src/` into a directory under `dist/`? Or is there processing involved? I would love to see an example using the modules, since lighter dependencies are one motivation for people to upgrade their dc.js.

We discussed folder layout in #1565. It sounds like a good idea but I want to see the details to understand how this preserves URLs.

The other two optional items would be nice - really up to you if you want to make the effort. I will be super relieved to be able to use arrow functions in examples - quite a few times I've checked them in without thinking, and had to go back later and fix them.

**Comment by kum-deepak** [comment]
> The only thing I see missing from the original plan is compatibility with D3v6. But I agree, it makes sense to release this first, and we can do point releases that update to each ES6 D3 module as it comes out, since those will also break user code. See #1545.
> 
> Maybe something like: 4.0.* is compatible with D3v5, while 4.1 uses the latest modules instead of the D3 bundle. Suggestions welcome!
> 
Depending on D3v6 production version, we may need changes in dc as well. Based on level of changes we may decide DCv4.1 or DCv5.

> I agree on all the non optional stuff. I am curious about the jsdoc updates, could you explain?
> 
We have ES6 classes and methods now. JSDOC should be able to figure that out without @method, @memberof, and @instance. Also, name of class needs to replace the earlier constructor method (lineChart -> LineChart). Also, we need to ensure that JSDOC is able to detect mixins (if needed with @mixin and @mixes). Some experimentation and then making changes. 

> Of the optional stuff, the two items regarding ES6 modules stand out to me. What does an ES6 module distribution look like - is it just a copy of `src/` into a directory under `dist/`? Or is there processing involved? I would love to see an example using the modules, since lighter dependencies are one motivation for people to upgrade their dc.js.
> 

Please see https://github.com/rollup/rollup/wiki/pkg.module. Need a bit of experimentation here. https://rollupjs.org/guide/en/#outputformat suggest `esm` as output format. It might very well be a copy of src in our case (as there is no transpilation).

Currently very few people use it, however, in all likelihood usage will grow.

> We discussed folder layout in #1565. It sounds like a good idea but I want to see the details to understand how this preserves URLs.
> 
There are multiple ways to achieve this. I will discuss in a separate PR/issue.

> The other two optional items would be nice - really up to you if you want to make the effort. I will be super relieved to be able to use arrow functions in examples - quite a few times I've checked them in without thinking, and had to go back later and fix them.

I am going to finish all the optional items as well. Only decision for each item is whether before or after release DCv4.

**Comment by gordonwoodhull** [comment]
> Depending on D3v6 production version, we may need changes in dc as well. Based on level of changes we may decide DCv4.1 or DCv5.

My impression is that D3v6 is going to be backward-compatible in its own interfaces. It's just that it will accept ES6 types such as iterators, and it uses ES6 internally. Since DCv4 requires ES6, we might be okay? They are currently annotating modules as being compatible with either the pre-ES6 version or the post-, see e.g. https://github.com/d3/d3-geo-projection/issues/177

I find semantic versioning a bit of a mind-bender. For example, what if you know you want to do a series of releases that will not be backward-compatible? It seems like from a strict reading of semver, you have to increment the major version number every time, or else hold onto everything until it's ready.

This is why I have used the minor version to mark breaking changes, and the major version to mark architectural shifts. Not great, I know. Glad for any guidance here.

> I am going to finish all the optional items as well. Only decision for each item is whether before or after release DCv4.

That's great. It's up to you. I don't see much risk in releasing first, although I think it will help users if we have one example using modules, so that they know what they are looking at. IMO we don't want to convert all example to use modules, just one to show the way.

**Comment by gordonwoodhull** [comment]
@kum-deepak, in your opinion, aside from #1575 and #1638, are we pretty much ready for a release of v4? (Those two issues are related, having to do with `this`)

I guess the other outstanding issue is the v6 changes to d3-selection and d3-transition. My understanding is that the big `d3.event` change has not percolated through to the behaviors like d3-zoom and d3-brush. That will be a big deal for us.

**Comment by kum-deepak** [comment]
Sorry @gordonwoodhull, was traveling last week. I will review and comment in a day :smile:

**Comment by kum-deepak** [comment]
Only one small item, I want to automate building module version of stock sample using a shell script. But, definitely not a release dependency.

**Comment by kum-deepak** [comment]
I have looked at the sample again. I will make manual changes - as I need to add / modify comments as well.

**Comment by kum-deepak** [comment]
No more pending items from my list :+1:

**Comment by gordonwoodhull** [comment]
OK! I guess we are ready for a release. 

I have some ongoing work to improve `elasticX`, which I will release as 3.2 and merge to es6. 

Then I'll do a careful review to make sure everything else is merged. (It's been messy doing development in two branches.)

Hopefully we can release 4.0 in the next week or two!

Thanks for all your efforts!

**Comment by gordonwoodhull** [comment]
There haven't been any bug reports since I released the beta almost a week ago.

Either people don't want to try the beta, or it is working correctly.

I will release as **4.0.0** in the next few days.
